File: backend/src/action-plans/action-plans.service.ts
import { Injectable, NotFoundException, ForbiddenException } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { CreateActionPlanDto } from './dto/create-action-plan.dto';
import { UpdateActionPlanDto } from './dto/update-action-plan.dto';
import { CreateActionItemDto } from './dto/create-action-item.dto';
import { UpdateActionItemDto } from './dto/update-action-item.dto';

@Injectable()
export class ActionPlansService {
  constructor(private prisma: PrismaService) {}

  async createActionPlan(
    userId: string,
    systemId: string,
    dto: CreateActionPlanDto,
  ) {
    return this.prisma.actionPlan.create({
      data: {
        title: dto.title,
        description: dto.description,
        status: dto.status || 'active',
        userId,
        systemId,
      },
      include: {
        actionItems: true,
      },
    });
  }

  async findAllActionPlans(userId: string, systemId: string) {
    return this.prisma.actionPlan.findMany({
      where: {
        userId,
        systemId,
      },
      include: {
        actionItems: {
          orderBy: {
            createdAt: 'asc',
          },
        },
      },
      orderBy: {
        createdAt: 'desc',
      },
    });
  }

  async findActionPlanById(id: string, userId: string, systemId: string) {
    const actionPlan = await this.prisma.actionPlan.findFirst({
      where: {
        id,
        userId,
        systemId,
      },
      include: {
        actionItems: {
          orderBy: {
            createdAt: 'asc',
          },
        },
      },
    });

    if (!actionPlan) {
      throw new NotFoundException('Action plan not found');
    }

    return actionPlan;
  }

  async updateActionPlan(
    id: string,
    userId: string,
    systemId: string,
    dto: UpdateActionPlanDto,
  ) {
    await this.findActionPlanById(id, userId, systemId);

    return this.prisma.actionPlan.update({
      where: { id },
      data: {
        title: dto.title,
        description: dto.description,
        status: dto.status,
      },
      include: {
        actionItems: true,
      },
    });
  }

  async deleteActionPlan(id: string, userId: string, systemId: string) {
    await this.findActionPlanById(id, userId, systemId);

    await this.prisma.actionPlan.delete({
      where: { id },
    });

    return { message: 'Action plan deleted successfully' };
  }

  async createActionItem(
    actionPlanId: string,
    userId: string,
    systemId: string,
    dto: CreateActionItemDto,
  ) {
    await this.findActionPlanById(actionPlanId, userId, systemId);

    return this.prisma.actionItem.create({
      data: {
        title: dto.title,
        description: dto.description,
        dueDate: dto.dueDate ? new Date(dto.dueDate) : null,
        priority: dto.priority || 'medium',
        actionPlanId,
      },
    });
  }

  async findActionItemsByPlanId(
    actionPlanId: string,
    userId: string,
    systemId: string,
  ) {
    await this.findActionPlanById(actionPlanId, userId, systemId);

    return this.prisma.actionItem.findMany({
      where: {
        actionPlanId,
      },
      orderBy: {
        createdAt: 'asc',
      },
    });
  }

  async findActionItemById(
    itemId: string,
    actionPlanId: string,
    userId: string,
    systemId: string,
  ) {
    await this.findActionPlanById(actionPlanId, userId, systemId);

    const actionItem = await this.prisma.actionItem.findFirst({
      where: {
        id: itemId,
        actionPlanId,
      },
    });

    if (!actionItem) {
      throw new NotFoundException('Action item not found');
    }

    return actionItem;
  }

  async updateActionItem(
    itemId: string,
    actionPlanId: string,
    userId: string,
    systemId: string,
    dto: UpdateActionItemDto,
  ) {
    await this.findActionItemById(itemId, actionPlanId, userId, systemId);

    return this.prisma.actionItem.update({
      where: { id: itemId },
      data: {
        title: dto.title,
        description: dto.description,
        dueDate: dto.dueDate ? new Date(dto.dueDate) : undefined,
        priority: dto.priority,
      },
    });
  }

  async completeActionItem(
    itemId: string,
    actionPlanId: string,
    userId: string,
    systemId: string,
  ) {
    await this.findActionItemById(itemId, actionPlanId, userId, systemId);

    return this.prisma.actionItem.update({
      where: { id: itemId },
      data: {
        status: 'completed',
        completedAt: new Date(),
      },
    });
  }

  async uncompleteActionItem(
    itemId: string,
    actionPlanId: string,
    userId: string,
    systemId: string,
  ) {
    await this.findActionItemById(itemId, actionPlanId, userId, systemId);

    return this.prisma.actionItem.update({
      where: { id: itemId },
      data: {
        status: 'pending',
        completedAt: null,
      },
    });
  }

  async deleteActionItem(
    itemId: string,
    actionPlanId: string,
    userId: string,
    systemId: string,
  ) {
    await this.findActionItemById(itemId, actionPlanId, userId, systemId);

    await this.prisma.actionItem.delete({
      where: { id: itemId },
    });

    return { message: 'Action item deleted successfully' };
  }
}

File: app/services/action_plans_service.py
import logging
from datetime import datetime
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select
from sqlalchemy.orm import selectinload
from fastapi import HTTPException, status

from app.models.action_plan import ActionPlan, ActionItem
from app.schemas.action_plan import (
    ActionPlanResponse,
    ActionItemResponse,
    CreateActionPlanRequest,
    UpdateActionPlanRequest,
    CreateActionItemRequest,
    UpdateActionItemRequest
)

logger = logging.getLogger(__name__)


class ActionPlansService:
    def __init__(self, db: AsyncSession):
        self.db = db

    async def create_action_plan(
        self,
        user_id: str,
        system_id: str,
        data: CreateActionPlanRequest
    ) -> ActionPlanResponse:
        action_plan = ActionPlan(
            user_id=user_id,
            system_id=system_id,
            title=data.title,
            description=data.description,
            status=data.status,
        )

        self.db.add(action_plan)
        await self.db.commit()
        await self.db.refresh(action_plan, ["action_items"])

        return ActionPlanResponse.from_orm_with_camel(action_plan)

    async def find_all_action_plans(self, user_id: str, system_id: str) -> list[ActionPlanResponse]:
        result = await self.db.execute(
            select(ActionPlan)
            .where(ActionPlan.user_id == user_id, ActionPlan.system_id == system_id)
            .options(selectinload(ActionPlan.action_items))
            .order_by(ActionPlan.created_at.desc())
        )
        action_plans = result.scalars().all()

        return [ActionPlanResponse.from_orm_with_camel(plan) for plan in action_plans]

    async def find_action_plan_by_id(self, plan_id: str, user_id: str, system_id: str) -> ActionPlanResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
            .options(selectinload(ActionPlan.action_items))
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        return ActionPlanResponse.from_orm_with_camel(action_plan)

    async def update_action_plan(
        self,
        plan_id: str,
        user_id: str,
        system_id: str,
        data: UpdateActionPlanRequest
    ) -> ActionPlanResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        if data.title is not None:
            action_plan.title = data.title
        if data.description is not None:
            action_plan.description = data.description
        if data.status is not None:
            action_plan.status = data.status

        await self.db.commit()
        await self.db.refresh(action_plan, ["action_items"])

        return ActionPlanResponse.from_orm_with_camel(action_plan)

    async def delete_action_plan(self, plan_id: str, user_id: str, system_id: str):
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        await self.db.delete(action_plan)
        await self.db.commit()

        return {"message": "Action plan deleted successfully"}

    async def create_action_item(
        self,
        plan_id: str,
        user_id: str,
        system_id: str,
        data: CreateActionItemRequest
    ) -> ActionItemResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        action_item = ActionItem(
            action_plan_id=plan_id,
            title=data.title,
            description=data.description,
            due_date=data.dueDate,
            priority=data.priority,
        )

        self.db.add(action_item)
        await self.db.commit()
        await self.db.refresh(action_item)

        return ActionItemResponse.from_orm_with_camel(action_item)

    async def find_action_items_by_plan_id(
        self,
        plan_id: str,
        user_id: str,
        system_id: str
    ) -> list[ActionItemResponse]:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem)
            .where(ActionItem.action_plan_id == plan_id)
            .order_by(ActionItem.created_at.desc())
        )
        action_items = result.scalars().all()

        return [ActionItemResponse.from_orm_with_camel(item) for item in action_items]

    async def find_action_item_by_id(
        self,
        item_id: str,
        plan_id: str,
        user_id: str,
        system_id: str
    ) -> ActionItemResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem).where(
                ActionItem.id == item_id,
                ActionItem.action_plan_id == plan_id
            )
        )
        action_item = result.scalar_one_or_none()

        if not action_item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action item not found"
            )

        return ActionItemResponse.from_orm_with_camel(action_item)

    async def update_action_item(
        self,
        item_id: str,
        plan_id: str,
        user_id: str,
        system_id: str,
        data: UpdateActionItemRequest
    ) -> ActionItemResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem).where(
                ActionItem.id == item_id,
                ActionItem.action_plan_id == plan_id
            )
        )
        action_item = result.scalar_one_or_none()

        if not action_item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action item not found"
            )

        if data.title is not None:
            action_item.title = data.title
        if data.description is not None:
            action_item.description = data.description
        if data.dueDate is not None:
            action_item.due_date = data.dueDate
        if data.priority is not None:
            action_item.priority = data.priority

        await self.db.commit()
        await self.db.refresh(action_item)

        return ActionItemResponse.from_orm_with_camel(action_item)

    async def complete_action_item(
        self,
        item_id: str,
        plan_id: str,
        user_id: str,
        system_id: str
    ) -> ActionItemResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem).where(
                ActionItem.id == item_id,
                ActionItem.action_plan_id == plan_id
            )
        )
        action_item = result.scalar_one_or_none()

        if not action_item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action item not found"
            )

        action_item.completed_at = datetime.utcnow()

        await self.db.commit()
        await self.db.refresh(action_item)

        return ActionItemResponse.from_orm_with_camel(action_item)

    async def uncomplete_action_item(
        self,
        item_id: str,
        plan_id: str,
        user_id: str,
        system_id: str
    ) -> ActionItemResponse:
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem).where(
                ActionItem.id == item_id,
                ActionItem.action_plan_id == plan_id
            )
        )
        action_item = result.scalar_one_or_none()

        if not action_item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action item not found"
            )

        action_item.completed_at = None

        await self.db.commit()
        await self.db.refresh(action_item)

        return ActionItemResponse.from_orm_with_camel(action_item)

    async def delete_action_item(self, item_id: str, plan_id: str, user_id: str, system_id: str):
        result = await self.db.execute(
            select(ActionPlan)
            .where(
                ActionPlan.id == plan_id,
                ActionPlan.user_id == user_id,
                ActionPlan.system_id == system_id
            )
        )
        action_plan = result.scalar_one_or_none()

        if not action_plan:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action plan not found"
            )

        result = await self.db.execute(
            select(ActionItem).where(
                ActionItem.id == item_id,
                ActionItem.action_plan_id == plan_id
            )
        )
        action_item = result.scalar_one_or_none()

        if not action_item:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Action item not found"
            )

        await self.db.delete(action_item)
        await self.db.commit()

        return {"message": "Action item deleted successfully"}
